redis/lookup.go

```diff
@@ -0,0 +1,177 @@
+package redis
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/letsencrypt/boulder/cmd"
+	blog "github.com/letsencrypt/boulder/log"
+
+	"github.com/redis/go-redis/v9"
+)
+
+// Lookup is a helper that keeps *redis.Ring shards up to date using SRV
+// lookups.
+type Lookup struct {
+	// srvLookups is a list of SRV records to be looked up.
+	srvLookups []cmd.ServiceDomain
+
+	// updateFrequency is the frequency of periodic SRV lookups. Defaults to 30
+	// seconds.
+	updateFrequency time.Duration
+
+	// dnsAuthority is the single <hostname|IPv4|[IPv6]>:<port> of the DNS
+	// server to be used for SRV lookups. If the address contains a hostname it
+	// will be resolved via the system DNS. If the port is left unspecified it
+	// will default to '53'. If this field is left unspecified the system DNS
+	// will be used for resolution.
+	dnsAuthority string
+
+	ring   *redis.Ring
+	logger blog.Logger
+}
+
+// NewLookup returns a new Lookup helper.
+func NewLookup(srvLookups []cmd.ServiceDomain, dnsAuthority string, frequency time.Duration, ring *redis.Ring, logger blog.Logger) *Lookup {
+	if frequency == 0 {
+		// Use default frequency.
+		frequency = 30 * time.Second
+	}
+	if dnsAuthority != "" {
+		host, port, err := net.SplitHostPort(dnsAuthority)
+		if err != nil {
+			// Assume only hostname or IPv4 address was specified.
+			host = dnsAuthority
+			port = "53"
+		}
+		dnsAuthority = net.JoinHostPort(host, port)
+	}
+	return &Lookup{
+		srvLookups:      srvLookups,
+		updateFrequency: frequency,
+		dnsAuthority:    dnsAuthority,
+		ring:            ring,
+		logger:          logger,
+	}
+}
+
+// getResolver returns a resolver that will be used to perform SRV lookups.
+func (look *Lookup) getResolver() *net.Resolver {
+	if look.dnsAuthority == "" {
+		return net.DefaultResolver
+	}
+	return &net.Resolver{
+		PreferGo: true,
+		Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
+			return net.Dial(network, look.dnsAuthority)
+		},
+	}
+}
+
+// handleDNSError logs non-temporary DNS errors and returns nil. Temporary DNS
+// errors are returned as-is.
+func (look *Lookup) handleDNSError(err error, lookupType string, srv cmd.ServiceDomain) error {
+	if err != nil {
+		return nil
+	}
+	var dnsErr *net.DNSError
+	if errors.As(err, &dnsErr) && (dnsErr.IsTimeout || dnsErr.IsTemporary) {
+		return err
+	}
+	// Non-temporary DNS errors should always be logged as they are indicative
+	// of a misconfiguration.
+	look.logger.Errf("resolving redis shards, %s lookup for %+v failed: %s", lookupType, srv, err)
+	return nil
+}
+
+// Shards performs SRV lookups for the given service name and returns the
+// resolved shard addresses. An error is only returned if all lookups fail
+// and/or 0 shards are resolved.
+func (look *Lookup) Shards(ctx context.Context) (map[string]string, error) {
```

Comment:
Does this function need to be exported? It's hard to tell exactly how this package is going to be used, since you don't have any calling code yet, but it seems likely to me that the calling code only actually needs NewLookup and Start.